### ISSUE TITLE
[Feedback?] Sending randomised batches of emails

### ISSUE THREAD

**Issue by S0larflare** [gophish/gophish#501]
First off, I never noticed the developer documentation, so gophish is easy enough to work out for someone who has never touched before (so kudos to Jordan there)

Anyway, for my usage, I wanted the ability to send batches of emails, delay sending and randomise the order (in order to avoid sending emails to people sat next to each other at the same time and spoiling the exercise). I've done this for myself now (which is an extension of the pull request that I've done) but wondered if there was any appetite for this for anyone else or if it is a bit more niche? If a few people want it I'll tidy it up a little and create a PR for it. Feedback welcomed.

![1](https://cloud.githubusercontent.com/assets/8939757/21393589/013caf3a-c78d-11e6-8bee-4741d2cf3175.png)
![2](https://cloud.githubusercontent.com/assets/8939757/21393588/013ae6e6-c78d-11e6-8e15-5c3f93e7c052.png)
![3](https://cloud.githubusercontent.com/assets/8939757/21393587/013a8fe8-c78d-11e6-9e65-131ad0876742.png)

**Comment by reynas** [comment]
+1

**Comment by jordan-wright** [comment]
I like the idea of randomized email sending, but I'm not sure this is the UI approach I would take.

There's a couple of ways we can go about it:

* Specifying the "batch size" and the delay between each batch
* Setting a start and end date and spreading out the sending over that range

In terms of simplicity in the UI, I feel like the second approach might be easier to understand and use. Basically we would be trickling the emails from the time the campaign starts to an end time.

What do y'all think?

**Comment by jordan-wright** [comment]
Also, sorry for the delay on this- holidays and all that happened 😄

**Comment by S0larflare** [comment]
Yes, I'd thought about heding that both ways as I was doing it by having it show the "end date" for when the last email would be sent. Working that out would be pretty easy, the problem was that I couldn't think of a nice way to actually present it on that screen, I wanted as little clutter as possible.

I can think of pros and cons for both systems. I think the reason I did it my way was a) it matched how I wanted to segregate batches for my own use of gophish and b) I assumed one of the biggest needs for this was for people who need to rate-limit themselves to avoid being caught in filters (only allowed 25 within 5 minutes, for example). More than happy to go with the popular opinion on this one though, would welcome the feedback.

Might need to check a few things on this with you anyway Jordan if it ends up moving towards being pu in to gophish.

**Comment by S0larflare** [comment]
Right,

I've updated this to take account of the changes you made (as of about a week ago, will need to check compatability with your latest commit but there isn't much to do on that front). I am having a couple of small issues though.

1. Have you minified the css? Is everything now served from /static/css/dist/gophish.css? I had added something in when I did this before the changes in order to have a 3 column section for the sending part, but that css file doesn't seem to be called any more.

2. When I came to thinking about it properly, it would be difficult to include something to say how long the campaign should last as it would mean knowing how many users are being targetted, which we might not know until they had been added...

3. There is an issue with how I have implimented the delay. I've made it simply pause the function for x seconds per group of y users. Performance wise this is fine as expected (I've tested it for a campaign with 800 users and never noticed a single difference) however there are timeouts using Gomail if I leave it too long (e.g 3 minutes per batch or every 10 minutes in total). I got around this quickly by restarting the connection before it would time out, but I'm not sure that this is the best way of handling it really. I'm wondering if it would be better handling that on the Gomail side perhaps?

**Comment by patilprateek** [comment]
Hi guys,

Relatively new having started using this tool, and liking the outputs that i have been getting from the campaigns created. To add to S0larflare's enhancement request, i am in favor of it and was trying to find options in the existing tool to implement. Need is like mentioned to randomizing the campaign by having a start & end date, and may be we can setup a time interval as to when the set of emails to be sent to the end users.

Not sure if i can just add a line of another query i have, it was in regards to having an option of attaching an image file directly in the sending template (the option i see now is of adding a source link from the web, where the image is hosted). Well this works, but partially as when the mails lands in the inbox outlooks blocks the link giving a message to download the contents in the top pane. Intention here is to make the campaign feel more genuine by putting in images that the users will feel coming from a trusted source.

**Comment by S0larflare** [comment]
You can add an image directly into the template pretty easily.

Add the image as an attached file for that email template, then using the exact name, reference it like this:

<img src="cid:your_image.png" />

**Comment by figjamalba** [comment]
Yes please!
 Randomised sending with a delay in batches would be great! The current super blast of mails tend to set off all the bells during a campaign.

**Comment by jordan-wright** [comment]
Just wanted to provide an update on this one. To implement this the "right way", I needed to refactor the email sending infrastructure. This will also have a few other benefits, like letting us fix #548 and #79 

To give some background on how this will work, I'm implementing a mail queue that's stored in the database. Each entry, or "maillog" will point to a campaign id, a result id, a time for "when to send" and a send count (how many times we've tried to send it).

This means that when we create a campaign, we can let users select a start and end date for sending emails. Then, Gophish can simply schedule each maillog's "when to send" field accordingly to spread out over that interval.

This is a big change and I want to make sure (as best I can 😄 ) that I don't break anything, so I'm taking my time with this one. Hopefully it will be ready in the next week or so, in which case I can then move to building the changes around the new mail infrastructure, like this issue.

**Comment by chrismaddalena** [comment]
This is terrific news. Will it be possible to specify the delay interval and how many batches to send? In other words, 20 email batches with a delay of 10m woild send 20 randomly selected emails every 10m?

**Comment by jordan-wright** [comment]
That's a good question, and it's 100% open for discussion!

It seems like there's two options: a "batch" option and a "duration" option. The batch option would be a bit more complex to implement both UI wise and backend wise and would allow users to, as you said, send x emails every y minutes.

The other option, the one I'm leaning towards right now, just has the user select a start and end date for sending emails. Then, we'll distribute the emails evenly over that period of time. It's less flexible, but should be quicker to implement.

What do you think?

**Comment by chrismaddalena** [comment]
The easier to implement option would most likely cover most use cases and help the most GoPhish users today. However, a batch-style sending mechanism would be helpful in some scenarios.

There are pros and cons to both. Any way to delay emails while keeping everything in one campaign (or as few campaigns as possible) is a significant help. For example, if the recipients are using Google Apps and Gmail, you have to send =<20 emails at a time with a delay of ~10-15m to safely avoid being blocked. Using the time window method, the GoPhish user can determine the appropriate time window for emails to be slowly sent out, but it could get tricky.

One "gotchya" might be how mail filtering handles a trickle of emails over a long period of time. I don't know if Google, and other mail filtering solutions, respond differently to email blasts every 10-15m vs. singular emails being slowly sent out over a few hours. There *might be* a difference between bursts of emails vs. a steady flow.

The biggest drawback I can think of from a social engineering point of view is you are often racing against time. Sometimes you want to hit everyone with the email all at once to raise the chance that one or more people will respond before someone else in the group raises an alarm. That's where a batch-style sending method works best, assuming you're also trying to defeat a mail filter. On other occasions, you may want to trickle the emails out to favor stealth and increase the chance that, should someone raise an alarm, the warning is forgotten by the time the email is received by someone else.

Looking at GoPhish's capabilities today, there is currently no way to trickle the emails out over time. There is a way to send batches of emails at various intervals if you create multiple campaigns and manually activate them at the right time. It occurs to me that anyone could easily script the activation of those campaigns via the API as well, so it's not quite as tedious as clicking the button yourself with a timer on your phone. Also, GoReport can combine the campaigns into one report, so the functionality  of batch emails *sort of* exists today, via a combination of things. Still, creating all the campaigns individually is tedious. Scriptable, perhaps, but tedious.

With that in mind, I feel the option you are leaning towards now, the time window, would be an excellent addition for the near future. Looking to the more distant future, having the option between both would be nice.

**Comment by S0larflare** [comment]
The issue with how I implemented it in my internal build (which is old) and how I had updated it for newer gophish versions traced back to it essentially being a hack on top of the exisiting code. Sending thousands of emails over a period of time worked, but it would take longer than the calculated time, for example. Retries could be set up (barring a weird bug with error handling), but the most effective way of fixing most email problems was to create a new dialer for each email, which was pretty wasteful. All this just made me realise that the way email is handled needed rearchitecturing, so I'm glad that's being tackled! :D

**Comment by magnusstubman** [comment]
What's the status on this issue? I'd *really* like to see this implemented!

**Comment by caseyagil** [comment]
Wanted to queue in on this, so I can follow this enhancement / issue.

**Comment by kjohnson5** [comment]
Will this be in the next release? This is something that would be beneficial within my organization.

**Comment by xtrasimplicity** [comment]
> The next step will be to add support for randomized sending instead of evenly spaced sending. This should be a piece of 🍰, but it might have to come after the next release, since I have a few features I want to knock out before drafting a new release in the next couple of days. Just know it's coming soon!

Hi @jordan-wright. Has this functionality been finalised? We trialed GoPhish a few years ago, but discovered that users were getting campaign emails in large batches, which resulted in alarm bells going off internally, which vastly impacted our campaign.

Keen to introduce GoPhish into our environment again, but looking for a way to avoid that particular issue re-occurring.

If it hasn't been finalised, is there an active issue/PR that I can contribute to? If not, if this is still on the roadmap, I'd be happy to work on a PR for this.

Thanks!
